### PR TITLE
Issue #4873 - fix timeout on ExecutorThreadPool.join()

### LINKED ISSUE #4873: Server.join not working when used with ExecutorThreadPool
**Jetty version**
9.4.28.v20200408
**Java version**
OpenJDK Runtime Environment (build 11.0.6+10-LTS) by Azul Systems, Inc.
**OS type/version**
MacOS Catalina 10.15.3
**Description**
Hi guys, 

`Server.join` is not working properly when given `ThreadPoolExecutor`. Basically, the implementation of `ExecutorThreadPool` calls `awaitTermination` with `stopTimeout` as opposed to `Integer.MAX_VALUE` in the previous versions and that's not really what should be done here. Documentation of `awaitTermination`: 
```
Blocks until all tasks have completed execution after a shutdown request, 
or the timeout occurs, or the current thread is interrupted, whichever happens first.
```
Basically, it doesn't care if the `shutdown` has been requested and will always wait at most `stopTimeout`.

Very simple reproduction example:
[prints "start" and "end" and exits immediately]
```
        ExecutorService executorService = Executors.newCachedThreadPool();
        Server server = new Server(new ExecutorThreadPool((ThreadPoolExecutor) executorService));
        server.setStopTimeout(0);
        server.start();
        System.out.println("start");
        server.join();
        System.out.println("end");
        System.exit(0);
```
vs 
[prints "start" and hangs forever]
```
        Server server = new Server();
        server.setStopTimeout(0);
        server.start();
        System.out.println("start");
        server.join();
        System.out.println("end");
        System.exit(0);
```

I think the simplest elegant fix from the Jetty side (we can also set `stopTimeout` to `Integer.MAX_VALUE` as a hack) would be to add logic into `ExecutorThreadPool.join` from `QueuedThreadPool`:
```
public void join() throws InterruptedException
    {
        synchronized (_joinLock)
        {
            while (isRunning())
            {
                _joinLock.wait();
            }
        }

        _executor.awaitTermination(getStopTimeout(), TimeUnit.MILLISECONDS);
    }
```
plus 
```
synchronized (_joinLock)
        {
            _joinLock.notifyAll();
        }
```
in `doStop`.

And thank you so much for such a great and very valuable project!

### PR DESCRIPTION
**Pull request by lachlan-roberts**
see issue #4873

### CODE REVIEW

**Comment by joakime** [comment]
Be mindful when you create the PR.
This was originally created to merge into `jetty-10.0.x`.
I've edited the "wants to merge" to point to `jetty-9.4.x`

File: jetty-util/src/main/java/org/eclipse/jetty/util/thread/ExecutorThreadPool.java [hunk 1]
```diff
@@ -302,7 +302,7 @@ protected void doStop() throws Exception
     @Override
     public void join() throws InterruptedException
     {
-        _executor.awaitTermination(getStopTimeout(), TimeUnit.MILLISECONDS);
+        _executor.awaitTermination(Long.MAX_VALUE, TimeUnit.MILLISECONDS);
```
Comment (joakime) [inline jetty-util/src/main/java/org/eclipse/jetty/util/thread/ExecutorThreadPool.java:305]: Why are we even doing `awaitTermination()`?

Isn't there a better choice?

Comment (joakime) [reply jetty-util/src/main/java/org/eclipse/jetty/util/thread/ExecutorThreadPool.java:305]: I guess a 292 million year wait isn't going to cause anybody any concerns.

Comment (lachlan-roberts) [reply jetty-util/src/main/java/org/eclipse/jetty/util/thread/ExecutorThreadPool.java:305]: Doesn't look like there is any better way to do it other than `awaitTermination()`.

It doesn't look like waiting only `stopTimeout` caused many concerns either, it has been like that for over 2 years and only one issue filed against it. I guess it just wasn't used much.

**Comment by sbordet** [comment]
@lachlan-roberts let's write a test case like reported in #4873 so that we ensure the behavior across refactorings.

**Review by sbordet — CHANGES_REQUESTED** [review]
@lachlan-roberts can we have the tests for any thread pool, not only the executor one?
Moving the test to the base class should be enough.

**Review by gregw — APPROVED** [review]

**Comment by lachlan-roberts** [comment]
@sbordet its not as simple as moving to the base class as the `AbstractThreadPoolTest` gives you a `SizedThreadPool` which does not implement `ContainerLifeCycle`. I have put the test there and now just return if the threadPool is not a `ContainerLifeCycle`.

**Pull request merged**